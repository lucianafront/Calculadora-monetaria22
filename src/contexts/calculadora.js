import axios from "axios";
import { useState, useEffect, createContext } from "react";

export const CalculadoraContext = createContext({});

function CalculadoraProvider({ children }) {
  const [currency, setCurrency] = useState([]);
  const [moeda, setMoeda] = useState("BRL");
  const [destino, setDestino] = useState("USD");
  const [valor, setValor] = useState(1);
  const [historico, setHistorico] = useState([]);
  const [historicoErros, setHistoricoErros] = useState([]);

 

  useEffect(() => {
    axios.get(`https://api.exchangerate-api.com/v4/latest/${moeda}`)
    .then((response) => {
      setCurrency(Object.keys(response.data.rates));
      setHistorico(JSON.parse(localStorage.getItem("historico")) || []);

    })
    .catch((erro)=>{
      gravarHistoricoLocalStorage(erro);

    })

  }, []);

  useEffect(() => {
  }, []);

  function converter() {
    axios.get(`https://api.exchangerate-api.com/v4/latest/${moeda}`)
    .then((response) => {
      const cotacao = response.data.rates;
      console.log(cotacao);
      const result = (valor * cotacao[destino]);
    
      
      gravarHistoricoLocalStorage(result);

    })
    .catch((error) => {
      gravarHistoricoErrosLocalStorage(error);
    });
  }



  function gravarHistoricoErrosLocalStorage(error)
  {
    let historicoErros = JSON.parse(localStorage.getItem("historicoErros")) || [];
    historicoErros.unshift({
      moeda: moeda,
      destino: destino,
      valor: valor,
      erro: error.message
    });
    setHistoricoErros(historicoErros);
    localStorage.setItem("historicoErros", JSON.stringify(historicoErros));
  }

  function gravarHistoricoLocalStorage(result)
  {
    let historico = JSON.parse(localStorage.getItem("historico")) || [];
    historico.unshift({
      moeda: moeda,
      destino: destino,
      valor: valor,
      resultado: result
    });

    setHistorico(historico);
    localStorage.setItem("historico", JSON.stringify(historico));
  }


  return (
    <CalculadoraContext.Provider value={{ currency, moeda, setDestino, setMoeda, valor, setValor, converter, historico, historicoErros  }}>
      {children}
    </CalculadoraContext.Provider>
  );
}

export default CalculadoraProvider;


