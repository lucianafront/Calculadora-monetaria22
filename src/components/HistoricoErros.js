import { useContext } from "react";
import { CalculadoraContext } from "../contexts/calculadora";

function HistoricoErros() {
    const { historicoErros } = useContext(CalculadoraContext);
    
    return (
        <div className="erros">
        <h1>Hist√≥rico de Erros: Qtd#{historicoErros.length}</h1>
        <ul>
            {historicoErros.map((item, index) => (
            <li key={index}>
                De {item.moeda} para {item.destino} - Valor: {item.valor} -
                Erro: {item.erro}
            </li>
            ))}
        </ul>
        </div>
    );
    }

export default HistoricoErros;