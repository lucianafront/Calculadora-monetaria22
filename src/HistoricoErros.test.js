import React from "react";
import { render } from "@testing-library/react";
import { CalculadoraContext } from "./contexts/calculadora";
import HistoricoErros from "./components/HistoricoErros";

describe("HistoricoErros", () => {
  test("renders component with historical errors", () => {
    const historicoErros = [
      {
        moeda: "USD",
        destino: "BRL",
        valor: 100,
        erro: "Invalid currency",
      },
      {
        moeda: "EUR",
        destino: "JPY",
        valor: 50,
        erro: "Conversion rate not available",
      },
    ];

    const { getByText } = render(
      <CalculadoraContext.Provider value={{ historicoErros}}>
      <HistoricoErros/>
    </CalculadoraContext.Provider>
    );

    expect(getByText("HistÃ³rico de Erros")).toBeInTheDocument();
    expect(getByText("De USD para BRL - Valor: 100 - Erro: Invalid currency")).toBeInTheDocument();
    expect(getByText("De EUR para JPY - Valor: 50 - Erro: Conversion rate not available")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const historicoErros = [
      {
        moeda: "USD",
        destino: "BRL",
        valor: 100,
        erro: "Invalid currency",
      },
      {
        moeda: "EUR",
        destino: "JPY",
        valor: 50,
        erro: "Conversion rate not available",
      },
    ];

    const { container } = render(
      <CalculadoraContext.Provider value={{ historicoErros}}>
      <HistoricoErros/>
    </CalculadoraContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });
});