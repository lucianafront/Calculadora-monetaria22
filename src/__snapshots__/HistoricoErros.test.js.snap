// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoricoErros matches snapshot 1`] = `
<div>
  <h1>
    Hist√≥rico de Erros
  </h1>
  <ul>
    <li>
      De 
      USD
       para 
      BRL
       - Valor: 
      100
       - Erro: 
      Invalid currency
    </li>
    <li>
      De 
      EUR
       para 
      JPY
       - Valor: 
      50
       - Erro: 
      Conversion rate not available
    </li>
  </ul>
</div>
`;
